<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Snakemake Workflows &mdash; Rockfish Cluster Software Guide 2.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="_static/sphinx-design.4cbf315f70debaebd550c87a6162cf0f.min.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
        <script src="_static/design-tabs.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Singularity container" href="Singularity.html" />
    <link rel="prev" title="Reproducibility Framework (RF)" href="RF.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/arch.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About ARCH</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="About_ARCH.html">About ARCH</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Resources on Rockfish</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="HPC_Terminology.html">ARCH Terminology</a></li>
<li class="toctree-l1"><a class="reference internal" href="Allocation.html">Allocation and Account Management</a></li>
<li class="toctree-l1"><a class="reference internal" href="Compute_Resources.html">Compute Node Summary</a></li>
<li class="toctree-l1"><a class="reference internal" href="Filesystems.html">Filesystems on Rockfish</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Connect and Transfer to Rockfish</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Connecting_to_Rockfish.html">Logging in to Rockfish</a></li>
<li class="toctree-l1"><a class="reference internal" href="File_Transfers.html">File Transfers</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Manage Software and Jobs</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Slurm.html">Job Scheduling with Slurm</a></li>
<li class="toctree-l1"><a class="reference internal" href="Software.html">Managing Softwares using Lmod</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Common Tasks and Workflows</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="OpenOnDemand.html">Open OnDemand</a></li>
<li class="toctree-l1"><a class="reference internal" href="Linux.html">Linux Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="RStudioServer.html">RStudio Server</a></li>
<li class="toctree-l1"><a class="reference internal" href="JupyterLab.html">JupyterLab</a></li>
<li class="toctree-l1"><a class="reference internal" href="Parallel.html">Parallel Computing Jobs</a></li>
<li class="toctree-l1"><a class="reference internal" href="Rpackages.html">Using R on the Rockfish</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">How to do Reproducible Research</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="RF.html">Reproducibility Framework (RF)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Snakemake Workflows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pipeline">Pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sra-toolkit">SRA Toolkit</a></li>
<li class="toctree-l2"><a class="reference internal" href="#cutadapt">Cutadapt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#cutadapt-snakemake-slurm-script">cutadapt snakemake.slurm.script</a></li>
<li class="toctree-l3"><a class="reference internal" href="#monitoring-submitted-jobs">Monitoring submitted jobs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#burrows-wheeler-alignment-tool">Burrows-Wheeler Alignment Tool</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#bwamem-snakemake-slurm-script">bwamem snakemake.slurm.script</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#remove-duplicates">Remove duplicates</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#rmdup-snakemake-slurm-script">rmdup snakemake.slurm.script</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#add-tags">Add tags</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#tags-snakemake-slurm-script">tags snakemake.slurm.script</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tabix">Tabix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#tabix-snakemake-slurm-script">tabix snakemake.slurm.script</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Singularity.html">Singularity container</a></li>
<li class="toctree-l1"><a class="reference internal" href="Virtual.html">Virtual Environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="Shorts.html">Short Tutorials</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bash_Quickstart.html">Bash Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="Managing_Group_Permissions.html">Manage Group Permissions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Support and Help</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Support.html">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="Service_Level_Expectations.html">Service Level Expectations</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Rockfish Cluster Software Guide</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Snakemake Workflows</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Snakemake.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="snakemake-workflows">
<h1>Snakemake Workflows<a class="headerlink" href="#snakemake-workflows" title="Permalink to this heading"></a></h1>
<a class="reference external image-reference" href="https://bioconda.github.io/recipes/snakemake/README.html"><img alt="https://img.shields.io/conda/dn/bioconda/snakemake.svg?label=Bioconda" src="https://img.shields.io/conda/dn/bioconda/snakemake.svg?label=Bioconda" /></a>
<a class="reference external image-reference" href="https://www.python.org"><img alt="https://img.shields.io/pypi/pyversions/snakemake.svg" src="https://img.shields.io/pypi/pyversions/snakemake.svg" /></a>
<a class="reference external image-reference" href="https://pypi.python.org/pypi/snakemake"><img alt="https://img.shields.io/pypi/v/snakemake.svg" src="https://img.shields.io/pypi/v/snakemake.svg" /></a>
<a class="reference external image-reference" href="https://hub.docker.com/r/snakemake/snakemake"><img alt="https://img.shields.io/github/workflow/status/snakemake/snakemake/Publish%20to%20Docker%20Hub?color=blue&amp;label=docker%20container&amp;branch=main" src="https://img.shields.io/github/workflow/status/snakemake/snakemake/Publish%20to%20Docker%20Hub?color=blue&amp;label=docker%20container&amp;branch=main" /></a>
<a class="reference external image-reference" href="https://github.com/snakemake/snakemake/actions?query=branch%3Amain+workflow%3ACI"><img alt="https://github.com/snakemake/snakemake/workflows/CI/badge.svg?branch=main&amp;label=tests" src="https://github.com/snakemake/snakemake/workflows/CI/badge.svg?branch=main&amp;label=tests" /></a>
<a class="reference external image-reference" href="https://stackoverflow.com/questions/tagged/snakemake"><img alt="https://img.shields.io/badge/stack-overflow-orange.svg" src="https://img.shields.io/badge/stack-overflow-orange.svg" /></a>
<p>The <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html">Snakemake</a> workflows management system is a tool to create reproducible and scalable data analyses.</p>
<p>The Snakemake language extends the Python language, adding syntactic structures for rule definition and additional controls. All added syntactic structures begin with a keyword followed by a code block that is either in the same line or indented and consists of multiple lines. The resulting syntax resembles that of original Python constructs.</p>
<p>A Snakemake workflow is defined by specifying rules in a Snakefile. The rules decompose the workflow into small steps by specifying how to create sets of output files from sets of input files. It will automatically determine the dependencies between the rules by matching file names.</p>
<p>This tutorial presents a bioinformatics pipeline using <code class="docutils literal notranslate"><span class="pre">Snakemake</span></code> and <a class="reference internal" href="RF.html#reproducibility-framework"><span class="std std-ref">the Reproducibility Framework (RF)</span></a>.</p>
<p>It will used the two classes of L1-associated somatic variants in the human brain from Salk Institute for Biological Studies dataset.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Bioproject: <code class="docutils literal notranslate"><span class="pre">PRJEB10849</span></code> SRA Study: <code class="docutils literal notranslate"><span class="pre">ERP012147</span></code>.</p>
<p><a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=ERR1016570">https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=ERR1016570</a></p>
</div>
<section id="pipeline">
<h2>Pipeline<a class="headerlink" href="#pipeline" title="Permalink to this heading"></a></h2>
<p>Then, let’s create the <a class="reference external" href="https://github.com/jhu-arch/arch-tutorial/tree/main/tutorial">pipeline</a> directory structure to store this tutorial.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 ~]$ </span>mkdir<span class="w"> </span>-p<span class="w"> </span>pipeline/_h
<span class="gp">[userid@login03 ~]$ </span>mkdir<span class="w"> </span>-p<span class="w"> </span>pipeline/cutadapt/_h
<span class="gp">[userid@login03 ~]$ </span>mkdir<span class="w"> </span>-p<span class="w"> </span>pipeline/cutadapt/bwamem/_h
<span class="gp">[userid@login03 ~]$ </span>mkdir<span class="w"> </span>-p<span class="w"> </span>pipeline/cutadapt/genome/_h
<span class="gp">[userid@login03 ~]$ </span>mkdir<span class="w"> </span>-p<span class="w"> </span>pipeline/cutadapt/bwamem/rmdup/_h
<span class="gp">[userid@login03 ~]$ </span>mkdir<span class="w"> </span>-p<span class="w"> </span>pipeline/cutadapt/bwamem/rmdup/tags/_h
<span class="gp">[userid@login03 ~]$ </span>mkdir<span class="w"> </span>-p<span class="w"> </span>pipeline/cutadapt/bwamem/rmdup/tags/tabix/_h
</pre></div>
</div>
<p>A directory structure like this is expected to run our pipeline.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 pipeline]$ </span>rf<span class="w"> </span>status
<span class="go">.                                   done  (step done)</span>
<span class="go">└── cutadapt                        done</span>
<span class="go">    ├── bwamem              ready to run  (step with the run file, but not performed)</span>
<span class="go">    │   └── rmdup          no run script</span>
<span class="go">    │       └── tags       no run script  (step without the run file)</span>
<span class="go">    │           └── tabix  no run script</span>
<span class="go">    └── genome                         no _h</span>
<span class="go">        └── hg19                ready to run</span>
<span class="go">            ├── bwa             ready to run</span>
<span class="go">            └── chromsizes      ready to run</span>
</pre></div>
</div>
<p>Once the scripts are ready to run, as is the case with the genome step. Just run the <code class="docutils literal notranslate"><span class="pre">rf</span></code> command recursively using the <code class="docutils literal notranslate"><span class="pre">-r</span></code> tag.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 pipeline]$ </span><span class="nb">cd</span><span class="w"> </span>genome/
<span class="gp">[userid@login03 genome]$ </span>rf<span class="w"> </span>sbatch<span class="w"> </span>-r<span class="w"> </span>.

<span class="go">Start /home/userid/pipeline/genome/hg19: 2022-05-04 17:30:53-04:00</span>
<span class="go">End /home/userid/pipeline/genome/hg19: 2022-05-04 17:30:53-04:00</span>
<span class="go">Start /home/userid/pipeline/genome/hg19/bwa: 2022-05-04 17:30:53-04:00</span>
<span class="go">End /home/userid/pipeline/genome/hg19/bwa: 2022-05-04 17:30:53-04:00</span>
<span class="go">Start /home/userid/pipeline/genome/hg19/chromsizes: 2022-05-04 17:30:53-04:00</span>
<span class="go">End /home/userid/pipeline/genome/hg19/chromsizes: 2022-05-04 17:30:53-04:00</span>

<span class="gp">[rdesouz4@login03 genome]$ </span>rf<span class="w"> </span>status
<span class="go">.                           no _h</span>
<span class="go">└── hg19                     done</span>
<span class="go">    ├── bwa                  done</span>
<span class="go">    └── chromsizes           done</span>
</pre></div>
</div>
</section>
<section id="sra-toolkit">
<h2>SRA Toolkit<a class="headerlink" href="#sra-toolkit" title="Permalink to this heading"></a></h2>
<p>To download sequence data files using SRA Toolkit, you need create a <code class="docutils literal notranslate"><span class="pre">run</span></code> file into <code class="docutils literal notranslate"><span class="pre">pipeline/_h</span></code> folder.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>!/bin/bash

<span class="gp">#</span>SBATCH<span class="w"> </span>-J<span class="w"> </span>sra_tools
<span class="gp">#</span>SBATCH<span class="w"> </span>-p<span class="w"> </span>defq
<span class="gp">#</span>SBATCH<span class="w"> </span>-N<span class="w"> </span><span class="m">1</span>
<span class="gp">#</span>SBATCH<span class="w"> </span>--time<span class="o">=</span><span class="m">2</span>:00:00
<span class="gp">#</span>login01SBATCH<span class="w"> </span>--cpus-per-task<span class="o">=</span><span class="m">4</span>
<span class="gp">#</span>SBATCH<span class="w"> </span>--output<span class="o">=</span>Array_test.%A_%a.out
<span class="gp">#</span>SBATCH<span class="w"> </span>--array<span class="o">=</span><span class="m">1</span>-101

<span class="go">ml sra-tools/3.0.0</span>

<span class="gp"># </span>Bioproject<span class="w"> </span>PRJEB10849<span class="w"> </span>samples

<span class="go">sra_numbers=($(echo {1016570..1016671}))</span>

<span class="go">sra_id=&#39;ERR&#39;${sra_numbers[ $SLURM_ARRAY_TASK_ID - 1 ]}</span>

<span class="go">prefetch --max-size 100G $sra_id --force yes --verify no</span>
<span class="go">fastq-dump --outdir . --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ${sra_id}/${sra_id}.sra</span>

<span class="go">rm $sra_id -Rf</span>
</pre></div>
</div>
<p>The  <code class="docutils literal notranslate"><span class="pre">rf</span></code> command will call the <code class="docutils literal notranslate"><span class="pre">run</span></code> script to retrieve SRA Normalized Format files with full base quality scores, and store them <code class="docutils literal notranslate"><span class="pre">fastq</span></code> files into <code class="docutils literal notranslate"><span class="pre">_m</span></code> folder.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 ~]$ </span><span class="nb">cd</span><span class="w"> </span>pipeline/
<span class="gp">[userid@login03 ~]$ </span>chmod<span class="w"> </span>+x<span class="w"> </span>_h/run
<span class="gp">[userid@login03 pipeline]$ </span>rf<span class="w"> </span>sbatch<span class="w"> </span>-v<span class="w"> </span>.
<span class="go">all: /home/userid/pipeline/_m/SUCCESS</span>

<span class="go">.ONESHELL:</span>
<span class="go">/home/userid/pipeline/_m/SUCCESS:</span>
<span class="go">      echo -n &quot;Start /home/userid/pipeline: &quot;; date --rfc-3339=seconds</span>
<span class="go">      mkdir /home/userid/pipeline/_m</span>
<span class="go">      cd /home/userid/pipeline/_m</span>
<span class="go">      sbatch ../_h/run &gt; nohup.out 2&gt;&amp;1</span>
<span class="go">      touch SUCCESS</span>
<span class="go">      echo -n &quot;End /home/userid/pipeline: &quot;; date --rfc-3339=seconds</span>

<span class="go">Start /home/userid/pipeline: 2022-04-27 16:14:52-04:00</span>
<span class="go">End /home/userid/pipeline: 2022-04-27 16:14:52-04:00</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><strong>Writing Workflows</strong> : “In Snakemake, <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/snakefiles/writing_snakefiles.html">workflows</a> are specified as Snakefiles. Inspired by GNU Make, a <a class="reference external" href="ttps://snakemake.readthedocs.io/en/stable/snakefiles/rules.html">Snakefile</a> contains rules that denote how to create output files from input files. Dependencies between rules are handled implicitly, by matching filenames of input files against output files. Thereby wildcards can be used to write general rules.”</p></li>
<li><p><strong>Snakefiles and Rules</strong> : “A Snakemake workflow defines a data analysis in terms of rules that are specified in the Snakefile.”</p></li>
</ul>
</div>
<p>We will create a hypothetical scenario with precedent steps, where for example the Level 5 (tabix) depends on the Level 4 (tags), and so on.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Level 1 (cutadapt)  -&gt;   Level 2 (bwamem) -&gt;   Level 3 (rmdup) -&gt;  Level 4 (tags) -&gt;  Level 5 (tabix)</strong></p>
</div>
</section>
<section id="cutadapt">
<h2>Cutadapt<a class="headerlink" href="#cutadapt" title="Permalink to this heading"></a></h2>
<img alt="" src="https://github.com/marcelm/cutadapt/workflows/CI/badge.svg" /><a class="reference external image-reference" href="https://pypi.python.org/pypi/cutadapt"><img alt="" src="https://img.shields.io/pypi/v/cutadapt.svg?branch=master" /></a>
<a class="reference external image-reference" href="https://codecov.io/gh/marcelm/cutadapt"><img alt="" src="https://codecov.io/gh/marcelm/cutadapt/branch/master/graph/badge.svg" /></a>
<a class="reference external image-reference" href="http://bioconda.github.io/recipes/cutadapt/README.html"><img alt="install with bioconda" src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" /></a>
<p>Cutadapt finds and removes adapter sequences, primers, poly-A tails and other types of unwanted sequence from your high-throughput sequencing reads. It helps with these trimming tasks by finding the adapter or primer sequences in an error-tolerant way.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 pipeline]$ </span><span class="nb">cd</span><span class="w"> </span>cutadapt/
<span class="gp">[userid@login03 cutadapt]$ </span>vi<span class="w"> </span>_h/run

<span class="gp">#</span>!/bin/bash

<span class="gp">#</span>SBATCH<span class="w"> </span>-J<span class="w"> </span>cutadapt
<span class="gp">#</span>SBATCH<span class="w"> </span>-p<span class="w"> </span>defq
<span class="gp">#</span>SBATCH<span class="w"> </span>--time<span class="o">=</span><span class="m">2</span>:00:00
<span class="gp">#</span>login01SBATCH<span class="w"> </span>--cpus-per-task<span class="o">=</span><span class="m">4</span>
<span class="gp">#</span>SBATCH<span class="w"> </span>--output<span class="o">=</span>cutadapt.job.%j.out

<span class="go">module load snakemake/7.6.0</span>

<span class="gp"># </span>Syntax<span class="w"> </span>to<span class="w"> </span>run<span class="w"> </span>it<span class="w"> </span>on<span class="w"> </span>Rockfish<span class="w"> </span>cluster
<span class="go">snakemake --jobs 101 --latency-wait 240 --cluster &#39;sbatch --parsable --distribution=arbitrary&#39; --snakefile ../_h/snakemake.slurm.script</span>
</pre></div>
</div>
<p>So, we need create a script to perform the rev_comp_seq. Given a DNA sequence in string object, it will return its reverse.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 cutadapt]$ </span>vi<span class="w"> </span>~/.local/bin/rc

<span class="gp">#</span>!/bin/bash
<span class="go">if [ ! -z &quot;$1&quot; ]; then</span>
<span class="go">    echo &quot;$1&quot; | tr &quot;[ATGCatgc]&quot; &quot;[TACGtacg]&quot; | rev</span>
<span class="go">else</span>
<span class="go">    echo &quot;&quot;</span>
<span class="go">    echo &quot;usage: rc DNASEQUENCE&quot;</span>
<span class="go">    echo &quot;&quot;</span>
<span class="go">fi</span>

<span class="gp">[userid@login03 cutadapt]$ </span>chmod<span class="w"> </span>+x<span class="w"> </span>~/.local/bin/rc
<span class="gp">[userid@login03 cutadapt]$ </span>vi<span class="w"> </span>_h/snakemake.slurm.script
</pre></div>
</div>
<section id="cutadapt-snakemake-slurm-script">
<h3>cutadapt snakemake.slurm.script<a class="headerlink" href="#cutadapt-snakemake-slurm-script" title="Permalink to this heading"></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os.path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>

<span class="n">SOURCE_DIR</span> <span class="o">=</span> <span class="s1">&#39;../../_m&#39;</span>
<span class="n">EXT</span> <span class="o">=</span> <span class="s1">&#39;_pass_1.fastq.gz&#39;</span>

<span class="k">def</span><span class="w"> </span><span class="nf">sample_dict_iter</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
  <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">iglob</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s1">&#39;/*&#39;</span><span class="o">+</span><span class="n">ext</span><span class="p">):</span>
      <span class="n">sample</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filename</span><span class="p">)[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ext</span><span class="p">)]</span>

      <span class="k">yield</span> <span class="n">sample</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;r1_in&#39;</span><span class="p">:</span> <span class="n">SOURCE_DIR</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;_pass_1.fastq.gz&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;r2_in&#39;</span><span class="p">:</span> <span class="n">SOURCE_DIR</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;_pass_2.fastq.gz&#39;</span>
        <span class="p">}</span>

<span class="n">SAMPLE_DICT</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">sample_dict_iter</span><span class="p">(</span><span class="n">SOURCE_DIR</span><span class="p">,</span> <span class="n">EXT</span><span class="p">)}</span>

<span class="c1">#insure errors propogate along pipe&#39;d shell commands</span>
<span class="n">shell</span><span class="o">.</span><span class="n">prefix</span><span class="p">(</span><span class="s2">&quot;set -o pipefail; &quot;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
  <span class="nb">input</span><span class="p">:</span>
      <span class="n">expand</span><span class="p">(</span><span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">_</span><span class="si">{suffix}</span><span class="s1">.fastq.gz&#39;</span><span class="p">,</span>
       <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLE_DICT</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>
       <span class="n">suffix</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">,</span><span class="s1">&#39;R2&#39;</span><span class="p">])</span>

<span class="n">rule</span> <span class="n">cutadapt</span><span class="p">:</span>
  <span class="nb">input</span><span class="p">:</span>
      <span class="n">r1</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">SAMPLE_DICT</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">sample</span><span class="p">][</span><span class="s1">&#39;r1_in&#39;</span><span class="p">],</span>
      <span class="n">r2</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">SAMPLE_DICT</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">sample</span><span class="p">][</span><span class="s1">&#39;r2_in&#39;</span><span class="p">]</span>
  <span class="n">output</span><span class="p">:</span>
      <span class="n">r1</span> <span class="o">=</span> <span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">_R1.fastq.gz&#39;</span><span class="p">,</span>
      <span class="n">r2</span> <span class="o">=</span> <span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">_R2.fastq.gz&#39;</span>

  <span class="n">params</span><span class="p">:</span>
      <span class="n">sample</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{sample}</span><span class="s1">&#39;</span>

  <span class="n">shell</span><span class="p">:</span>
<span class="w">      </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">  module load cutadapt/3.2</span>

<span class="sd">  export PATH=$HOME&#39;/.local/bin:&#39;$PATH</span>

<span class="sd">  R1_ADAPTER=&#39;AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT&#39;</span>
<span class="sd">  R2_ADAPTER=&#39;CAAGCAGAAGACGGCATACGAGANNNNNNNGTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT&#39;</span>

<span class="sd">  NESTED_PRIMER=&#39;TAACTAACCTGCACAATGTGCAC&#39;</span>

<span class="sd">  R1_FRONT=${{R1_ADAPTER}}</span>
<span class="sd">  R2_FRONT=${{R2_ADAPTER}}${{NESTED_PRIMER}}</span>
<span class="sd">  R1_END=`rc ${{R2_FRONT}}`</span>
<span class="sd">  R2_END=`rc ${{R1_FRONT}}`</span>

<span class="sd">  QUALITY_BASE=33</span>
<span class="sd">  QUALITY_CUTOFF=28</span>
<span class="sd">  MINIMUM_LENGTH=36</span>
<span class="sd">  ADAPTOR_OVERLAP=5</span>
<span class="sd">  ADAPTOR_TIMES=4</span>

<span class="sd">  cutadapt -j 0 --quality-base=${{QUALITY_BASE}} --quality-cutoff=${{QUALITY_CUTOFF}} --minimum-length=${{MINIMUM_LENGTH}} --overlap=${{ADAPTOR_OVERLAP}} --times=${{ADAPTOR_TIMES}} --front=${{R1_FRONT}} --adapter=${{R1_END}} --paired-output tmp.2.{params.sample}.fastq -o tmp.1.{params.sample}.fastq {input.r1} {input.r2} &gt; {params.sample}_R1.cutadapt.out</span>

<span class="sd">  cutadapt -j 0 --quality-base=${{QUALITY_BASE}} --quality-cutoff=${{QUALITY_CUTOFF}} --minimum-length=${{MINIMUM_LENGTH}} --overlap=${{ADAPTOR_OVERLAP}} --times=${{ADAPTOR_TIMES}} --front=${{R2_FRONT}} --adapter=${{R2_END}} --paired-output {output.r1} -o {output.r2} tmp.2.{params.sample}.fastq tmp.1.{params.sample}.fastq &gt; {params.sample}_R2.cutadapt.out</span>

<span class="sd">  rm -f tmp.2.{params.sample}.fastq tmp.1.{params.sample}.fastq</span>

<span class="sd">&#39;&#39;&#39;</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 cutadapt]$ </span>chmod<span class="w"> </span>+x<span class="w"> </span>_h/run
<span class="gp">[userid@login03 cutadapt]$ </span>rf<span class="w"> </span>sbatch<span class="w"> </span>.

<span class="go">Start /home/userid/pipeline/cutadapt: 2022-05-04 14:35:06-04:00</span>
<span class="go">End /home/userid/pipeline/cutadapt: 2022-05-04 14:35:06-04:00</span>
</pre></div>
</div>
</section>
<section id="monitoring-submitted-jobs">
<h3>Monitoring submitted jobs<a class="headerlink" href="#monitoring-submitted-jobs" title="Permalink to this heading"></a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login02 _m]$ </span>sqme
<span class="go">  USER   ACCOUNT        JOBID PARTITION       NAME NODES  CPUS TIME_LIMIT     TIME NODELIST ST REASON</span>
<span class="go">userid   rfadmin      4157118 defq      snakejob.c     1     1    1:00:00    21:15     c221 R None</span>
<span class="go">userid   rfadmin      4157146 defq      snakejob.c     1     1    1:00:00    21:15     c301 R None</span>
<span class="go">userid   rfadmin      4157061 defq      snakejob.c     1     1    1:00:00    21:26     c157 R None</span>
<span class="go">userid   rfadmin      4157072 defq      snakejob.c     1     1    1:00:00    21:26     c132 R None</span>
<span class="go">userid   rfadmin      4157102 defq      snakejob.c     1     1    1:00:00    21:26     c303 R None</span>
<span class="go">userid   rfadmin      4157046 defq        cutadapt     1     1    2:00:00    21:28     c124 R None</span>
</pre></div>
</div>
<p>To monitoring all submitted processed jobs, <code class="docutils literal notranslate"><span class="pre">tail</span> <span class="pre">-f</span></code> on the file called <code class="docutils literal notranslate"><span class="pre">cutadapt.job.&lt;JOBID&gt;.out</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 cutadapt]$ </span>cat<span class="w"> </span>_m/cutadapt.job.4157046.out

<span class="go">Building DAG of jobs...</span>
<span class="go">Using shell: /usr/bin/bash</span>
<span class="go">Provided cluster nodes: 200</span>
<span class="go">Job stats:</span>
<span class="go">job         count    min threads    max threads</span>
<span class="go">--------  -------  -------------  -------------</span>
<span class="go">all             1              1              1</span>
<span class="go">cutadapt      101              1              1</span>
<span class="go">total         102              1              1</span>

<span class="go">Select jobs to execute...</span>

<span class="go">[Wed May  4 14:48:20 2022]</span>
<span class="go">rule cutadapt:</span>
<span class="go">    input: ../../_m/ERR1016599_pass_1.fastq.gz, ../../_m/ERR1016599_pass_2.fastq.gz</span>
<span class="go">    output: ../_m/ERR1016599_R1.fastq.gz, ../_m/ERR1016599_R2.fastq.gz</span>
<span class="go">    jobid: 26</span>
<span class="go">    wildcards: sample=ERR1016599</span>
<span class="go">    resources: mem_mb=1709, disk_mb=1709, tmpdir=/tmp</span>

<span class="go">Submitted job 26 with external jobid &#39;4157048&#39;.</span>

<span class="go">[Wed May  4 14:48:20 2022]</span>
<span class="go">rule cutadapt:</span>
<span class="go">    input: ../../_m/ERR1016661_pass_1.fastq.gz, ../../_m/ERR1016661_pass_2.fastq.gz</span>
<span class="go">    output: ../_m/ERR1016661_R1.fastq.gz, ../_m/ERR1016661_R2.fastq.gz</span>
<span class="go">    jobid: 86</span>
<span class="go">    wildcards: sample=ERR1016661</span>
<span class="go">    resources: mem_mb=3245, disk_mb=3245, tmpdir=/tmp</span>

<span class="go">........</span>
<span class="go">........</span>
<span class="go">........</span>
<span class="go">........</span>

<span class="go">[Wed May  4 14:48:30 2022]</span>
<span class="go">rule cutadapt:</span>
<span class="go">    input: ../../_m/ERR1016581_pass_1.fastq.gz, ../../_m/ERR1016581_pass_2.fastq.gz</span>
<span class="go">    output: ../_m/ERR1016581_R1.fastq.gz, ../_m/ERR1016581_R2.fastq.gz</span>
<span class="go">    jobid: 85</span>
<span class="go">    wildcards: sample=ERR1016581</span>
<span class="go">    resources: mem_mb=1891, disk_mb=1891, tmpdir=/tmp</span>

<span class="go">Submitted job 85 with external jobid &#39;4157148&#39;.</span>
<span class="go">[Wed May  4 14:49:33 2022]</span>
<span class="go">Finished job 37.</span>
<span class="go">1 of 102 steps (1%) done</span>
<span class="go">[Wed May  4 14:50:31 2022]</span>
<span class="go">Finished job 30.</span>
<span class="go">2 of 102 steps (2%) done</span>
<span class="go">[Wed May  4 14:51:35 2022]</span>
<span class="go">Finished job 16.</span>
<span class="go">3 of 102 steps (3%) done</span>
<span class="go">[Wed May  4 14:51:48 2022]</span>
<span class="go">Finished job 25.</span>
<span class="go">4 of 102 steps (4%) done</span>
<span class="go">[Wed May  4 14:51:49 2022]</span>
<span class="go">Finished job 87.</span>
<span class="go">5 of 102 steps (5%) done</span>
</pre></div>
</div>
<p>Also, it is possible to see the outputs for each sample processed, just monitoring the file called <code class="docutils literal notranslate"><span class="pre">slurm-&lt;snakejobid&gt;.out</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login02 _m]$ </span>cat<span class="w"> </span>slurm-4157147.out

<span class="go">Building DAG of jobs...</span>
<span class="go">Using shell: /usr/bin/bash</span>
<span class="go">Provided cores: 1 (use --cores to define parallelism)</span>
<span class="go">Rules claiming more threads will be scaled down.</span>
<span class="go">Select jobs to execute...</span>

<span class="go">[Wed May  4 14:48:37 2022]</span>
<span class="go">rule cutadapt:</span>
<span class="go">    input: ../../_m/ERR1016667_pass_1.fastq.gz, ../../_m/ERR1016667_pass_2.fastq.gz</span>
<span class="go">    output: ../_m/ERR1016667_R1.fastq.gz, ../_m/ERR1016667_R2.fastq.gz</span>
<span class="go">    jobid: 0</span>
<span class="go">    wildcards: sample=ERR1016667</span>
<span class="go">    resources: mem_mb=1000, disk_mb=1000, tmpdir=/tmp</span>

<span class="go">[Wed May  4 14:51:25 2022]</span>
<span class="go">Finished job 0.</span>
<span class="go">1 of 1 steps (100%) done</span>
</pre></div>
</div>
</section>
</section>
<section id="burrows-wheeler-alignment-tool">
<h2>Burrows-Wheeler Alignment Tool<a class="headerlink" href="#burrows-wheeler-alignment-tool" title="Permalink to this heading"></a></h2>
<a class="reference external image-reference" href="https://github.com/lh3/bwa/actions"><img alt="Build Status" src="https://github.com/lh3/bwa/actions/workflows/ci.yaml/badge.svg" /></a>
<a class="reference external image-reference" href="https://sourceforge.net/projects/bio-bwa/files/?source=navbar"><img alt="SourceForge Downloads" src="https://img.shields.io/sourceforge/dt/bio-bwa.svg" /></a>
<a class="reference external image-reference" href="https://github.com/lh3/bwa/releases"><img alt="GitHub Downloads" src="https://img.shields.io/github/downloads/lh3/bwa/total.svg" /></a>
<a class="reference external image-reference" href="https://anaconda.org/bioconda/bwa"><img alt="BioConda Install" src="https://img.shields.io/conda/dn/bioconda/bwa.svg" /></a>
<p><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a> is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome. It consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 cutadapt]$ </span><span class="nb">cd</span><span class="w"> </span>bwamem
<span class="gp">[userid@login03 bwamem]$ </span>vi<span class="w"> </span>_h/run
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>

<span class="c1">#SBATCH -J bwamem</span>
<span class="c1">#SBATCH -p defq</span>
<span class="c1">#SBATCH --time=2:00:00</span>
<span class="c1">#login01SBATCH --cpus-per-task=4</span>
<span class="c1">#SBATCH --output=bwamem.job.job.%j.out</span>

<span class="n">module</span> <span class="n">load</span> <span class="n">snakemake</span><span class="o">/</span><span class="mf">7.6.0</span>

<span class="c1"># Syntax to run it on Rockfish cluster</span>
<span class="n">snakemake</span> <span class="o">--</span><span class="n">jobs</span> <span class="mi">101</span> <span class="o">--</span><span class="n">latency</span><span class="o">-</span><span class="n">wait</span> <span class="mi">240</span> <span class="o">--</span><span class="n">cluster</span> <span class="s1">&#39;sbatch --parsable --distribution=arbitrary&#39;</span> <span class="o">--</span><span class="n">snakefile</span> <span class="o">../</span><span class="n">_h</span><span class="o">/</span><span class="n">snakemake</span><span class="o">.</span><span class="n">slurm</span><span class="o">.</span><span class="n">script</span>
</pre></div>
</div>
<section id="bwamem-snakemake-slurm-script">
<h3>bwamem snakemake.slurm.script<a class="headerlink" href="#bwamem-snakemake-slurm-script" title="Permalink to this heading"></a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 bwamem]$ </span>vi<span class="w"> </span>_h/snakemake.slurm.script
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os.path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>

<span class="n">SOURCE_DIR</span> <span class="o">=</span> <span class="s1">&#39;../../_m&#39;</span>
<span class="n">EXT</span> <span class="o">=</span> <span class="s1">&#39;_R1.fastq.gz&#39;</span>

<span class="k">def</span><span class="w"> </span><span class="nf">sample_dict_iter</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">iglob</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s1">&#39;/*&#39;</span><span class="o">+</span><span class="n">ext</span><span class="p">):</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filename</span><span class="p">)[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ext</span><span class="p">)]</span>
        <span class="k">yield</span> <span class="n">sample</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;r1_in&#39;</span><span class="p">:</span> <span class="n">SOURCE_DIR</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;_R1.fastq.gz&#39;</span><span class="p">,</span>
                                 <span class="s1">&#39;r2_in&#39;</span><span class="p">:</span> <span class="n">SOURCE_DIR</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;_R2.fastq.gz&#39;</span>
                    <span class="p">}</span>

<span class="n">SAMPLE_DICT</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">sample_dict_iter</span><span class="p">(</span><span class="n">SOURCE_DIR</span><span class="p">,</span> <span class="n">EXT</span><span class="p">)}</span>

<span class="c1">#insure errors propogate along pipe&#39;d shell commands</span>
<span class="n">shell</span><span class="o">.</span><span class="n">prefix</span><span class="p">(</span><span class="s2">&quot;set -o pipefail; &quot;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span><span class="p">,</span>
             <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLE_DICT</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">rule</span> <span class="n">bwamem</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">r1</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">SAMPLE_DICT</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">sample</span><span class="p">][</span><span class="s1">&#39;r1_in&#39;</span><span class="p">],</span>
            <span class="n">r2</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">SAMPLE_DICT</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">sample</span><span class="p">][</span><span class="s1">&#39;r2_in&#39;</span><span class="p">]</span>

    <span class="n">output</span><span class="p">:</span>
        <span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span>

    <span class="n">params</span><span class="p">:</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{sample}</span><span class="s1">&#39;</span>

    <span class="n">shell</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    module load bwa-mem/0.7.17 samtools/1.15.1</span>

<span class="sd">    export PATH=$HOME&#39;/.local/bin:&#39;$PATH</span>

<span class="sd">    GENOME=&#39;../../../../genome/bwa/_m/hg19.fa&#39;</span>

<span class="sd">    bwa mem -T 19 -t 4 ${{GENOME}} {input.r1} {input.r2} 2&gt; {params.sample}.stderr | samtools view -S -b - &gt; {output}</span>
<span class="sd">&#39;&#39;&#39;</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 bwamem]$ </span>chmod<span class="w"> </span>+x<span class="w"> </span>_h/run
<span class="gp">[userid@login03 bwamem]$ </span>rf<span class="w"> </span>sbatch<span class="w"> </span>.
</pre></div>
</div>
</section>
</section>
<section id="remove-duplicates">
<h2>Remove duplicates<a class="headerlink" href="#remove-duplicates" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://github.com/apuapaquola/slavseq_rf/blob/master/pipeline/fastq/cutadapt/bwamem/rmdup/_h/slavseq_rmdup.pl">rmdup</a> is a script part of the SLAV-Seq protocol written by Apuã Paquola, coded in Perl to read .bam input files and apply samtools software to treat paired-end reads and single-end reads.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 cutadapt]$ </span><span class="nb">cd</span><span class="w"> </span>rmdup
<span class="gp">[userid@login03 rmdup]$ </span>vi<span class="w"> </span>_h/run
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>

<span class="c1">#SBATCH -J rmdup</span>
<span class="c1">#SBATCH -p defq</span>
<span class="c1">#SBATCH --time=2:00:00</span>
<span class="c1">#SBATCH --cpus-per-task=4</span>
<span class="c1">#SBATCH --output=rmdup.job.job.%j.out</span>

<span class="n">module</span> <span class="n">load</span> <span class="n">snakemake</span><span class="o">/</span><span class="mf">7.6.0</span>

<span class="c1"># Syntax to run it on Rockfish cluster</span>
<span class="n">snakemake</span> <span class="o">--</span><span class="n">jobs</span> <span class="mi">101</span> <span class="o">--</span><span class="n">latency</span><span class="o">-</span><span class="n">wait</span> <span class="mi">240</span> <span class="o">--</span><span class="n">cluster</span> <span class="s1">&#39;sbatch --parsable --distribution=arbitrary&#39;</span> <span class="o">--</span><span class="n">snakefile</span> <span class="o">../</span><span class="n">_h</span><span class="o">/</span><span class="n">snakemake</span><span class="o">.</span><span class="n">slurm</span><span class="o">.</span><span class="n">script</span>
</pre></div>
</div>
<section id="rmdup-snakemake-slurm-script">
<h3>rmdup snakemake.slurm.script<a class="headerlink" href="#rmdup-snakemake-slurm-script" title="Permalink to this heading"></a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 rmdup]$ </span>vi<span class="w"> </span>_h/snakemake.slurm.script
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os.path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>

<span class="n">SOURCE_DIR</span> <span class="o">=</span> <span class="s1">&#39;../../_m&#39;</span>
<span class="n">EXT</span> <span class="o">=</span> <span class="s1">&#39;.bam&#39;</span>

<span class="k">def</span><span class="w"> </span><span class="nf">sample_dict_iter</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">iglob</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s1">&#39;/*&#39;</span><span class="o">+</span><span class="n">ext</span><span class="p">):</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filename</span><span class="p">)[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ext</span><span class="p">)]</span>
        <span class="k">yield</span> <span class="n">sample</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;filename&#39;</span><span class="p">:</span> <span class="n">filename</span><span class="p">}</span>


<span class="n">SAMPLE_DICT</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">sample_dict_iter</span><span class="p">(</span><span class="n">SOURCE_DIR</span><span class="p">,</span> <span class="n">EXT</span><span class="p">)}</span>

<span class="c1">#insure errors propogate along pipe&#39;d shell commands</span>
<span class="n">shell</span><span class="o">.</span><span class="n">prefix</span><span class="p">(</span><span class="s2">&quot;set -o pipefail; &quot;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLE_DICT</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">rule</span> <span class="n">process_one_sample</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">SAMPLE_DICT</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">sample</span><span class="p">][</span><span class="s1">&#39;filename&#39;</span><span class="p">]</span>

    <span class="n">output</span><span class="p">:</span>
        <span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span>
    <span class="n">log</span><span class="p">:</span>
        <span class="n">stderr</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{sample}</span><span class="s1">.stderr&#39;</span><span class="p">,</span>
        <span class="n">stdout</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{sample}</span><span class="s1">.stdout&#39;</span>
    <span class="n">shell</span><span class="p">:</span>
        <span class="s1">&#39;../_h/slavseq_rmdup.pl </span><span class="si">{input}</span><span class="s1"> </span><span class="si">{output}</span><span class="s1">&#39;</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 rmdup]$ </span>chmod<span class="w"> </span>+x<span class="w"> </span>_h/run
<span class="gp">[userid@login03 rmdup]$ </span>rf<span class="w"> </span>sbatch<span class="w"> </span>.
</pre></div>
</div>
</section>
</section>
<section id="add-tags">
<h2>Add tags<a class="headerlink" href="#add-tags" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://github.com/apuapaquola/slavseq_rf/blob/master/pipeline/fastq/cutadapt/bwamem/rmdup/tags/_h/add_tags.pl">tags</a> is a script part of the SLAV-Seq protocol written by Apuã Paquola, coded in Perl to add the custom flags into bam files.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 rmdup]$ </span><span class="nb">cd</span><span class="w"> </span>tags
<span class="gp">[userid@login03 tags]$ </span>vi<span class="w"> </span>_h/run
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>!/bin/bash

<span class="gp">#</span>SBATCH<span class="w"> </span>-J<span class="w"> </span>tags
<span class="gp">#</span>SBATCH<span class="w"> </span>-p<span class="w"> </span>defq
<span class="gp">#</span>SBATCH<span class="w"> </span>--time<span class="o">=</span><span class="m">2</span>:00:00
<span class="gp">#</span>login01SBATCH<span class="w"> </span>--cpus-per-task<span class="o">=</span><span class="m">4</span>
<span class="gp">#</span>SBATCH<span class="w"> </span>--output<span class="o">=</span>tags.job.job.%j.out

<span class="go">module load snakemake/7.6.0</span>

<span class="gp"># </span>Syntax<span class="w"> </span>to<span class="w"> </span>run<span class="w"> </span>it<span class="w"> </span>on<span class="w"> </span>Rockfish<span class="w"> </span>cluster
<span class="go">snakemake --jobs 101 --latency-wait 240 --cluster &#39;sbatch --parsable --distribution=arbitrary&#39; --snakefile ../_h/snakemake.slurm.script</span>
</pre></div>
</div>
<section id="tags-snakemake-slurm-script">
<h3>tags snakemake.slurm.script<a class="headerlink" href="#tags-snakemake-slurm-script" title="Permalink to this heading"></a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 tags]$ </span>vi<span class="w"> </span>_h/snakemake.slurm.script
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os.path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>

<span class="n">SOURCE_DIR</span> <span class="o">=</span> <span class="s1">&#39;../../_m&#39;</span>
<span class="n">EXT</span> <span class="o">=</span> <span class="s1">&#39;.bam&#39;</span>

<span class="k">def</span><span class="w"> </span><span class="nf">sample_dict_iter</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">iglob</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s1">&#39;/*&#39;</span><span class="o">+</span><span class="n">ext</span><span class="p">):</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filename</span><span class="p">)[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ext</span><span class="p">)]</span>
        <span class="k">yield</span> <span class="n">sample</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;filename&#39;</span><span class="p">:</span> <span class="n">SOURCE_DIR</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;.bam&#39;</span><span class="p">}</span>


<span class="n">SAMPLE_DICT</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">sample_dict_iter</span><span class="p">(</span><span class="n">SOURCE_DIR</span><span class="p">,</span> <span class="n">EXT</span><span class="p">)}</span>

<span class="c1">#insure errors propogate along pipe&#39;d shell commands</span>
<span class="n">shell</span><span class="o">.</span><span class="n">prefix</span><span class="p">(</span><span class="s2">&quot;set -o pipefail; &quot;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span><span class="p">,</span>
               <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLE_DICT</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">rule</span> <span class="n">tags</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s1">&#39;../../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span>

    <span class="n">output</span><span class="p">:</span>
        <span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span>

    <span class="n">params</span><span class="p">:</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{sample}</span><span class="s1">&#39;</span>

    <span class="n">shell</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    module load samtools/1.15.1</span>

<span class="sd">    export PERL5LIB=$HOME&#39;/perl5/lib/perl5/&#39;</span>
<span class="sd">    export CONSENSUS=&#39;ATGTACCCTAAAACTTAGAGTATAATAAA&#39;</span>
<span class="sd">    export PATH=$HOME&#39;/.local/bin:&#39;$PATH</span>

<span class="sd">    GENOME=&#39;../../../../../../genome/bwa/_m/hg19.fa&#39;</span>

<span class="sd">    PREFIX_LENGTH=`perl -e &#39;print length($ENV{{CONSENSUS}})+2&#39;`</span>
<span class="sd">    R1_FLANK_LENGTH=750</span>
<span class="sd">    R2_FLANK_LENGTH=${{PREFIX_LENGTH}}</span>
<span class="sd">    SOFT_CLIP_LENGTH_THRESHOLD=5</span>

<span class="sd">    (samtools view -h {input} | ../_h/add_tags_hts.pl --genome_fasta_file ${{GENOME}} --prefix_length ${{PREFIX_LENGTH}} --consensus ${{CONSENSUS}} --r1_flank_length ${{R1_FLANK_LENGTH}} --r2_flank_length ${{R2_FLANK_LENGTH}} --soft_clip_length_threshold ${{SOFT_CLIP_LENGTH_THRESHOLD}} | samtools view -S -b - &gt; {output}) 2&gt; {params.sample}.stderr</span>
<span class="sd">&#39;&#39;&#39;</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 tags]$ </span>chmod<span class="w"> </span>+x<span class="w"> </span>_h/run
<span class="gp">[userid@login03 tags]$ </span>rf<span class="w"> </span>sbatch<span class="w"> </span>.
</pre></div>
</div>
</section>
</section>
<section id="tabix">
<h2>Tabix<a class="headerlink" href="#tabix" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="http://www.htslib.org/doc/tabix.html">Tabix</a> indexes a TAB-delimited genome position file in.tab.bgz and creates an index file (in.tab.bgz.tbi or in.tab.bgz.csi) when region is absent from the command-line.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 tags]$ </span><span class="nb">cd</span><span class="w"> </span>tabix
<span class="gp">[userid@login03 tabix]$ </span>vi<span class="w"> </span>_h/run
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>!/bin/bash

<span class="gp">#</span>SBATCH<span class="w"> </span>-J<span class="w"> </span>tabix
<span class="gp">#</span>SBATCH<span class="w"> </span>-p<span class="w"> </span>defq
<span class="gp">#</span>SBATCH<span class="w"> </span>--time<span class="o">=</span><span class="m">2</span>:00:00
<span class="gp">#</span>login01SBATCH<span class="w"> </span>--cpus-per-task<span class="o">=</span><span class="m">4</span>
<span class="gp">#</span>SBATCH<span class="w"> </span>--output<span class="o">=</span>tabix.job.job.%j.out

<span class="go">module load snakemake/7.6.0</span>

<span class="gp"># </span>Syntax<span class="w"> </span>to<span class="w"> </span>run<span class="w"> </span>it<span class="w"> </span>on<span class="w"> </span>Rockfish<span class="w"> </span>cluster
<span class="go">snakemake --jobs 101 --latency-wait 240 --cluster &#39;sbatch --parsable --distribution=arbitrary&#39; --snakefile ../_h/snakemake.slurm.script</span>
</pre></div>
</div>
<section id="tabix-snakemake-slurm-script">
<h3>tabix snakemake.slurm.script<a class="headerlink" href="#tabix-snakemake-slurm-script" title="Permalink to this heading"></a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 tabix]$ </span>vi<span class="w"> </span>_h/snakemake.slurm.script
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os.path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span>

<span class="n">SOURCE_DIR</span> <span class="o">=</span> <span class="s1">&#39;../../_m&#39;</span>
<span class="n">EXT</span> <span class="o">=</span> <span class="s1">&#39;.bam&#39;</span>

<span class="k">def</span><span class="w"> </span><span class="nf">sample_dict_iter</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">iglob</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s1">&#39;/*&#39;</span><span class="o">+</span><span class="n">ext</span><span class="p">):</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filename</span><span class="p">)[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ext</span><span class="p">)]</span>
        <span class="k">yield</span> <span class="n">sample</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;filename&#39;</span><span class="p">:</span> <span class="n">SOURCE_DIR</span> <span class="o">+</span> <span class="s1">&#39;/&#39;</span> <span class="o">+</span> <span class="n">sample</span> <span class="o">+</span> <span class="s1">&#39;.bam&#39;</span><span class="p">}</span>

<span class="n">SAMPLE_DICT</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">sample_dict_iter</span><span class="p">(</span><span class="n">SOURCE_DIR</span><span class="p">,</span> <span class="n">EXT</span><span class="p">)}</span>

<span class="c1">#insure errors propogate along pipe&#39;d shell commands</span>
<span class="n">shell</span><span class="o">.</span><span class="n">prefix</span><span class="p">(</span><span class="s2">&quot;set -o pipefail; &quot;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">expand</span><span class="p">(</span><span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.</span><span class="si">{ext}</span><span class="s1">&#39;</span><span class="p">,</span>
               <span class="n">sample</span><span class="o">=</span><span class="n">SAMPLE_DICT</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>
             <span class="n">ext</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;bgz&#39;</span><span class="p">,</span> <span class="s1">&#39;bgz.tbi&#39;</span><span class="p">])</span>

<span class="n">rule</span> <span class="n">tabix</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="s1">&#39;../../_m/</span><span class="si">{sample}</span><span class="s1">.bam&#39;</span>

    <span class="n">output</span><span class="p">:</span>
        <span class="n">bgz</span> <span class="o">=</span> <span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bgz&#39;</span><span class="p">,</span>
        <span class="n">tbi</span> <span class="o">=</span> <span class="s1">&#39;../_m/</span><span class="si">{sample}</span><span class="s1">.bgz.tbi&#39;</span>

    <span class="n">params</span><span class="p">:</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{sample}</span><span class="s1">&#39;</span>

    <span class="n">shell</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    module load tabix/1.13 samtools/1.15.1 bzip2/1.0.8</span>

<span class="sd">    export PATH=$HOME&#39;/.local/bin:&#39;$PATH</span>

<span class="sd">    TMP_DIR=&#39;tmp.{params.sample}&#39;</span>
<span class="sd">    mkdir ${{TMP_DIR}}</span>

<span class="sd">    export LC_ALL=C</span>

<span class="sd">    ( samtools view {input} | ../_h/sam_to_tabix.py 2&gt;{params.sample}.stderr | sort --temporary-directory=${{TMP_DIR}} --buffer-size=10G -k1,1 -k2,2n -k3,3n | bgzip2 -c &gt; {output.bgz} )</span>

<span class="sd">    rmdir ${{TMP_DIR}}</span>

<span class="sd">    tabix -s 1 -b 2 -e 3 -0 {output.bgz}</span>
<span class="sd">&#39;&#39;&#39;</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 tabix]$ </span>chmod<span class="w"> </span>+x<span class="w"> </span>_h/run
<span class="gp">[userid@login03 tabix]$ </span>rf<span class="w"> </span>sbatch<span class="w"> </span>.
</pre></div>
</div>
<p>Once you coded the pipeline, just run <a class="reference internal" href="RF.html#reproducibility-framework"><span class="std std-ref">the Reproducibility Framework (RF)</span></a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">├── pipeline                          no _h</span>
<span class="go">  ├── cutadapt                 ready to run</span>
<span class="go">  │    └── bwamem             no run script</span>
<span class="go">  │        └── rmdup          no run script</span>
<span class="go">  │          └── tags         no run script</span>
<span class="go">  │               └── tabix   no run script</span>
<span class="go">  └── genome                          no _h</span>
<span class="go">      └── hg19                 ready to run</span>
<span class="go">          ├── bwa              ready to run</span>
<span class="go">          └── chromsizes       ready to run</span>
</pre></div>
</div>
<p>You run one level at a time, or you can use the <code class="docutils literal notranslate"><span class="pre">-r</span></code> option for recursive. It will perform the <code class="docutils literal notranslate"><span class="pre">rf</span></code> command, once the level 1 is finishes, it will run next level, so consecutively.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">[userid@login03 ~]$ </span>interact<span class="w"> </span>-c<span class="w"> </span><span class="m">2</span><span class="w"> </span>-t<span class="w"> </span><span class="m">120</span>
<span class="gp">[userid@c010 ~]$ </span><span class="nb">cd</span><span class="w"> </span>pipeline
<span class="gp">[userid@c010 ~]$ </span>rf<span class="w"> </span>run<span class="w"> </span>-r<span class="w"> </span>.
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The <code class="docutils literal notranslate"><span class="pre">rf</span></code> command is validated to run in interactive mode, so far.</p>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="RF.html" class="btn btn-neutral float-left" title="Reproducibility Framework (RF)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Singularity.html" class="btn btn-neutral float-right" title="Singularity container" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, How-to on the Rockfish Cluster.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>