# First load the module
#        module load snakemake
#
# usage: $ snakemake -np ../sorted_reads/B.bam
#          snakemake -np ../mapped_reads/A.bam ../mapped_reads/B.bam
#          snakemake -np ../mapped_reads/{A,B}.bam


rule bwa_map:
    input:
        "../data/genome.fa",
        "../data/samples/{sample}.fastq"
    output:
        "../mapped_reads/{sample}.bam"
    shell:
        '''
    module load bwa-mem samtools
    bwa mem {input} | samtools view -Sb - > {output}
'''
